### PR TITLE
feat(manager/gomod): capture indirect dependencies for Go Modules

### LINKED ISSUE #19416: Capture indirect dependencies for Go Modules
### What would you like Renovate to be able to do?

To be able to have insight into the versions of transitive dependencies available through Go modules through Renovate's APIs, as well as longer-term possibly even updating transitive dependencies as part of Renovate, it'd be handy to parse the `go.sum` files.

### If you have any ideas on how this should be implemented, please tell us here.

Introduce a new `manager` for `go.sum`.

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by jamietanna**
## Changes

To allow folks to be able to discover the dependencies available through
indirect (transitive) dependencies for Go modules, we can parse them as
a separate `depType`.

This also makes sure that it's clear that `indirect` dependencies do not
get updated by Renovate.

## Context

- Closes #19416

As noted in #19416, this provides the ability to discover indirect dependencies.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/gomod/extract.ts [hunk 1]
```diff
@@ -70,9 +70,13 @@ export function extractPackageFile(content: string): PackageFile | null {
...
-      if (requireMatch && !line.endsWith('// indirect')) {
-        logger.trace({ lineNumber }, `require line: "${line}"`);
-        const dep = getDep(lineNumber, requireMatch, 'require');
+      if (requireMatch) {
+        const requireType = line.endsWith('// indirect')
+          ? 'indirect'
+          : 'require';
+
+        logger.trace({ lineNumber }, `${requireType} line: "${line}"`);
+        const dep = getDep(lineNumber, requireMatch, requireType);
```
Comment (viceice) [inline lib/modules/manager/gomod/extract.ts:79]: set `enabled=false` flag, so we can later use package rules to override by user

Comment (jamietanna) [reply lib/modules/manager/gomod/extract.ts:79]: Or are you thinking to add that in as a new field?

Comment (jamietanna) [reply lib/modules/manager/gomod/extract.ts:79]: Maybe something like?

```typescript
          dep.managerData!.enabled = false

```

Comment (viceice) [reply lib/modules/manager/gomod/extract.ts:79]: no, it should be added to `Package<T>` interface

Comment (jamietanna) [reply lib/modules/manager/gomod/extract.ts:79]: Sure! Happy for it to be in this PR or shall I do a separate one?

Comment (viceice) [reply lib/modules/manager/gomod/extract.ts:79]: this one

File: lib/modules/manager/gomod/extract.ts [hunk 2]
```diff
@@ -70,10 +70,17 @@ export function extractPackageFile(content: string): PackageFile | null {
...
       const requireMatch = regEx(/^require\s+([^\s]+)\s+([^\s]+)/).exec(line);
-      if (requireMatch && !line.endsWith('// indirect')) {
-        logger.trace({ lineNumber }, `require line: "${line}"`);
-        const dep = getDep(lineNumber, requireMatch, 'require');
-        deps.push(dep);
+      if (requireMatch) {
+        if (line.endsWith('// indirect')) {
+          logger.trace({ lineNumber }, `indirect line: "${line}"`);
+          const dep = getDep(lineNumber, requireMatch, 'indirect');
+          dep.enabled = false;
```
Comment (viceice) [inline lib/modules/manager/gomod/extract.ts:77]:
```suggestion
          dep.enabled = false;
          dep.skipReason = 'disabled';
```

so until sombody modifies `updateDependency` to support `indirect` deps, nobody should be able to enable via package rule

File: lib/modules/manager/gomod/update.spec.ts [hunk 3]
```diff
@@ -362,5 +362,16 @@ describe('modules/manager/gomod/update', () => {
       });
       expect(res).toBeNull();
     });
+
+    it('should return null for indirect upgrades', () => {
+      const upgrade = {
+        depName: 'github.com/davecgh/go-spew',
+        managerData: { lineNumber: 5 }, // TODO
```
Comment (viceice) [inline lib/modules/manager/gomod/update.spec.ts:369]: Why todo? Add a description.

**Review by viceice — COMMENTED** [review]
Should this PR really close the issue? as this only extracts indirect deps, but doesn't allow to update them

**Comment by jamietanna** [comment]
Thanks for the comments, some good things I'd missed 👍 

> Should this PR really close the issue?

Yes, right now, I'm purely looking at getting the data for these dependencies, but I'm happy following up with #19430 to actually update them 👍

File: lib/modules/manager/gomod/extract.ts (hunk 2 above)
Comment (rarkins) [reply lib/modules/manager/gomod/extract.ts:77]: Shall we fix that at the same time? I'm assuming @jamietanna wants to update too

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
Replaced by #19431

**Pull request closed without merging**